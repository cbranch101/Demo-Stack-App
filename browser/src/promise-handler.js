import fetchMock from "fetch-mock"

const getDeferred = () => {
    const deferred = {}
    const promise = new Promise((resolve, reject) => {
        deferred.resolve = resolve
        deferred.reject = reject
    })
    deferred.promise = promise
    return deferred
}

const delayPromise = amount => new Promise(res => setTimeout(res, amount))

export const mockResponses = mockOptions => {
    const queue = [getDeferred()]
    const urlIndex = {}
    const updatedOptions = mockOptions.map(mockOption => {
        const { delay: delayAmount = 0, ...toPassAlong } = mockOption
        return {
            ...toPassAlong,
            response: (url, ...opts) => {
                const handleResponse = () => {
                    const deferred = queue.shift()
                    deferred.url = url
                    const responses = !Array.isArray(mockOption.response)
                        ? [mockOption.response]
                        : mockOption.response
                    const urlKey = `${url}_${JSON.stringify(opts)}`
                    urlIndex[urlKey] = urlIndex[urlKey] || 0
                    const currentIndex = urlIndex[urlKey]
                    const response = responses[currentIndex]
                    const returnValue =
                        typeof response === "function" ? response(url, ...opts) : response
                    const resolver = () => {
                        if ("MOCK_ERROR" in returnValue) {
                            deferred.reject(returnValue["MOCK_ERROR"])
                        } else {
                            deferred.resolve(returnValue)
                        }
                    }
                    if (mockOption.chain) {
                        mockOption.chain(Promise.resolve(returnValue)).then(resolver)
                    } else {
                        resolver()
                    }
                    if (Array.isArray(mockOption.response)) {
                        urlIndex[urlKey] = urlIndex[urlKey] + 1
                    }
                    queue.push(getDeferred())
                    return response
                }

                return delayAmount > 0
                    ? delayPromise(delayAmount).then(handleResponse)
                    : handleResponse()
            }
        }
    })

    updatedOptions.forEach(option => {
        fetchMock.mock(option)
    })
    return {
        next: () => {
            return queue[queue.length - 1].promise
        },
        unmock: () => fetchMock.restore()
    }
}
