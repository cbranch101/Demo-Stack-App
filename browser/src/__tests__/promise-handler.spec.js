import { mockResponses } from "../promise-handler"

const unpackResponse = promise => promise.then(response => response.json())

test("Should correctly mock array response", async () => {
    const listener = mockResponses([
        {
            matcher: "/users",
            method: "GET",
            response: [{ name: "John" }, { name: "Tom" }]
        }
    ])

    const firstResponse = await unpackResponse(fetch("/users"))
    expect(firstResponse).toMatchSnapshot()
    const secondResponse = await unpackResponse(fetch("/users"))
    expect(secondResponse).toMatchSnapshot()
    listener.unmock()
})

test("Should allow for an array of responses to be supplied and return a listener", async () => {
    const listener = mockResponses([
        {
            matcher: "/users",
            method: "GET",
            response: [{ name: "John" }, { name: "Tom" }]
        }
    ])

    setTimeout(() => fetch("/users"), 1)
    setTimeout(() => fetch("/users"), 2)
    const firstReturn = await listener.next()
    expect(firstReturn).toMatchSnapshot()
    const secondReturn = await listener.next()
    expect(secondReturn).toMatchSnapshot()
    listener.unmock()
})

test("Should handle a single response correctly", async () => {
    const listener = mockResponses([
        {
            matcher: "/users",
            method: "GET",
            response: { name: "John" }
        }
    ])

    setTimeout(() => fetch("/users"), 1)
    const firstReturn = await listener.next()
    expect(firstReturn).toMatchSnapshot()
    listener.unmock()
})

test("Should handle a function responses correctly", async () => {
    const listener = mockResponses([
        {
            matcher: "/users",
            method: "GET",
            response: url => ({
                name: url
            })
        }
    ])

    setTimeout(() => fetch("/users"), 1)
    const firstReturn = await listener.next()
    expect(firstReturn).toMatchSnapshot()
    listener.unmock()
})

test("Should be possible to throw errors", async () => {
    const listener = mockResponses([
        {
            matcher: "/users",
            method: "GET",
            response: { MOCK_ERROR: new Error("message") }
        }
    ])

    setTimeout(() => fetch("/users"), 1)
    try {
        await listener.next()
    } catch (e) {
        expect(e.message).toEqual("message")
        listener.unmock()
    }
})

test("Should correctly handle mocked requests to the same url", async () => {
    const listener = mockResponses([
        {
            matcher: "/users",
            method: "GET",
            response: [{ name: "John" }, { name: "Tom" }]
        },
        {
            matcher: "/users",
            method: "POST",
            response: { name: "George" }
        }
    ])

    setTimeout(() => fetch("/users"), 1)
    setTimeout(() => fetch("/users", { method: "POST" }), 2)
    setTimeout(() => fetch("/users"), 3)
    const firstReturn = await listener.next()
    expect(firstReturn).toMatchSnapshot()
    const secondReturn = await listener.next()
    expect(secondReturn).toMatchSnapshot()
    const thirdReturn = await listener.next()
    expect(thirdReturn).toMatchSnapshot()
    listener.unmock()
})

test("Should be able to handle for processing steps after fetch", async () => {
    let finished = false
    const listener = mockResponses([
        {
            chain: promise => promise.then(data => data),
            matcher: "/users",
            method: "GET",
            response: [{ name: "John" }, { name: "Tom" }]
        }
    ])

    setTimeout(
        () =>
            fetch("/users").then(data => {
                finished = true
                return data
            }),
        1
    )
    await listener.next()
    expect(finished).toEqual(true)

    listener.unmock()
})
