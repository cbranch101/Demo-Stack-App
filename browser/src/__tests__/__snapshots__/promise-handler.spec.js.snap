// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should allow for an array of responses to be supplied and return a listener 1`] = `
Object {
  "name": "John",
}
`;

exports[`Should allow for an array of responses to be supplied and return a listener 2`] = `
Object {
  "name": "Tom",
}
`;

exports[`Should correctly handle mocked requests to the same url 1`] = `
Object {
  "name": "John",
}
`;

exports[`Should correctly handle mocked requests to the same url 2`] = `
Object {
  "name": "George",
}
`;

exports[`Should correctly handle mocked requests to the same url 3`] = `
Object {
  "name": "Tom",
}
`;

exports[`Should correctly mock array response 1`] = `
Object {
  "name": "John",
}
`;

exports[`Should correctly mock array response 2`] = `
Object {
  "name": "Tom",
}
`;

exports[`Should handle a function responses correctly 1`] = `
Object {
  "name": "/users",
}
`;

exports[`Should handle a single response correctly 1`] = `
Object {
  "name": "John",
}
`;
